model_compression_toolkit/pytorch/graph_substitutions/substitutions/multi_head_attention_decomposition.py

```diff
@@ -0,0 +1,718 @@
+                                            output_shape=params.output_shape,
+                                            weights={},
+                                            layer_class=torch.transpose,
+                                            op_call_args=[1, 2],
+                                            op_call_kwargs={},
+                                            functional_op=torch.transpose)
+        graph.add_node_with_in_edges(transpose_out_node, [proj_out_node])
+        return transpose_out_node
+
+    @staticmethod
+    def _connect_to_graph(graph, mha_node, q_node, k_node, v_node, output_permute_node):
+        """
+        connect subgraph to input graph
+        Args:
+            graph: input graph
+            mha_node: MHA node to substitute inputs and outputs with
+            q_permute_node: 1st input to MHA node
+            k_permute_node: 2nd input to MHA node
+            v_permute_node: 3rd input to MHA node
+            output_permute_node: output node of MHA node
+
+        Returns:
+
+        """
+        query_in_edge, key_in_edge, value_in_edge = graph.in_edges(mha_node)
+        graph.add_edge(query_in_edge[0], q_node, **graph.get_edge_data(*query_in_edge, 0))
+        graph.add_edge(key_in_edge[0], k_node, **graph.get_edge_data(*key_in_edge, 0))
+        graph.add_edge(value_in_edge[0], v_node, **graph.get_edge_data(*value_in_edge, 0))
+        graph.remove_edge(query_in_edge[0], mha_node)
+        graph.remove_edge(key_in_edge[0], mha_node)
+        graph.remove_edge(value_in_edge[0], mha_node)
+        graph.reconnect_out_edges(current_node=mha_node, new_node=output_permute_node)
+
+    def substitute(self,
+                   graph: Graph,
+                   mha_node: BaseNode) -> Graph:
+
+        if mha_node.reuse:
+            raise Exception("MCT doesn't support reuse of MultiHeadAttention layer")
+        params = MHAParams(mha_node)
+
+        # check for MHA Parameters
+
+        # Only batch first network is supported
+        if BATCH_FIRST in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[BATCH_FIRST] is not True:
+                raise Exception('Only batch first network is supported, please fix your network')
+        else:
+            raise Exception('Only batch first network is supported, please fix your network')
+
+        # Add Zero Attn feature is Not Implemented
+        if ADD_ZERO_ATTN in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[ADD_ZERO_ATTN] is not False:
+                raise Exception('Add Zero Attn feature is Not Implemented')
+
+        # Add Bias KV feature is Not Implemented
+        if ADD_BIAS_KV in mha_node.framework_attr.keys():
+            if mha_node.framework_attr[ADD_BIAS_KV] is not False:
+                raise Exception('Add Bias KV feature is Not Implemented')
+
+        # Dropout is Not Implemented
+        if DROPOUT in mha_node.framework_attr:
```

Comment:
no need for this limitation: we only run inference mode, so dropout is an identity